weave/trace/weave_client.py

```diff
@@ -1970,34 +1975,190 @@ def _op_runs(self, op_def: Op) -> Sequence[Call]:
+                prev_call_processor_jobs = current_call_processor_jobs
+
+                # Sleep briefly to allow background threads to make progress
+                time.sleep(refresh_interval)
+
+        # Do the actual flush
         self._flush()
 
+        # Final callback with no pending jobs
+        if callback is not None:
+            final_status = WeaveFlushStatus(
+                main_jobs=0,
+                fastlane_jobs=0,
+                call_processor_jobs=0,
+                total_jobs=0,
+                completed_since_last_update=0,
+                total_completed=total_completed,
+                max_total_jobs=max_total_jobs,
+                has_pending_jobs=False,
+            )
+            callback(final_status)
+
     def _flush(self) -> None:
-        # Used to wait until all currently enqueued jobs are processed
+        """Used to wait until all currently enqueued jobs are processed."""
         if not self.future_executor._in_thread_context.get():
             self.future_executor.flush()
         if self.future_executor_fastlane:
             self.future_executor_fastlane.flush()
         if self._server_is_flushable:
-            # We don't want to do an instance check here because it could
-            # be susceptible to shutdown race conditions. So we save a boolean
-            # _server_is_flushable and only call this if we know the server is
-            # flushable. The # type: ignore is safe because we check the type
-            # first.
-            self.server.call_processor.wait_until_all_processed()  # type: ignore
+            self.server.call_processor.stop_accepting_new_work_and_flush_queue()  # type: ignore
 
-    def _send_file_create(self, req: FileCreateReq) -> Future[FileCreateRes]:
+    def _get_pending_jobs(self) -> tuple[int, int, int, int]:
+        """Get the current number of pending jobs for each type.
+
+        Returns:
+            tuple[int, int, int, int]: A tuple containing:
+                - main_jobs: Number of pending jobs in the main executor
+                - fastlane_jobs: Number of pending jobs in the fastlane executor
+                - call_processor_jobs: Number of pending jobs in the call processor
+                - total_jobs: Total number of pending jobs
+        """
+        main_jobs = self.future_executor.num_outstanding_futures
+        fastlane_jobs = 0
         if self.future_executor_fastlane:
-            # If we have a separate upload worker pool, use it
-            return self.future_executor_fastlane.defer(self.server.file_create, req)
-        return self.future_executor.defer(self.server.file_create, req)
+            fastlane_jobs = self.future_executor_fastlane.num_outstanding_futures
+        call_processor_jobs = 0
+        if self._server_is_flushable:
+            call_processor_jobs = self.server.call_processor.num_outstanding_jobs  # type: ignore
+
+        total_jobs = main_jobs + fastlane_jobs + call_processor_jobs
```

Comment:
nit: "total jobs" seems a little necessary

Reply:
i'm going to keep it in the typedDict, removing this local var though